network/src/application/interface.rs

```diff
@@ -15,7 +16,11 @@ use std::collections::{hash_map::Entry, HashMap};
 /// to the application as well as a specific `Sender` for cloning across threads and sending requests.
 #[async_trait]
 pub trait NetworkInterface {
+    /// A copyable sender for sending outbound messages
     type Sender;
```

Comment:
if the `NetworkInterface` doesn't require anything about the `Sender`, is it really providing an interface? if I took a generic `NetworkInterface` type, what do I do with this `Sender`? right now it claims to do nothing.

even though it's not a trait, today's sender supports

```rust
impl<TMessage: Message> NetworkSender<TMessage> {
    pub async fn dial_peer(
        &mut self,
        peer: PeerId,
        addr: NetworkAddress,
    ) -> Result<(), NetworkError>;

    pub async fn disconnect_peer(&mut self, peer: PeerId) -> Result<(), NetworkError>;

    pub fn send_to(
        &mut self,
        recipient: PeerId,
        protocol: ProtocolId,
        message: TMessage,
    ) -> Result<(), NetworkError>;

    pub fn send_to_many(
        &mut self,
        recipients: impl Iterator<Item = PeerId>,
        protocol: ProtocolId,
        message: TMessage,
    ) -> Result<(), NetworkError>;

    pub async fn send_rpc(
        &mut self,
        recipient: PeerId,
        protocol: ProtocolId,
        req_msg: TMessage,
        timeout: Duration,
    ) -> Result<TMessage, RpcError>;
}
```

i guess I would expect the `NetworkInterface` to provide something similar, where you can send a "serializable" message over a protocol to a PeerNetworkId

Reply:
Well, here's the reason:

We actually wrap these `NetworkSender`s right now in custom `ApplicationNetworkSender` per application.  Usually, they provide minimal add to even do that.

I can go back and patch the `NetworkSender` in instead of the custom `NetworkSender`s which is more of the bespoke coding problems in the network interfacing.

Just one thing at a time :/

Reply:
I'll come back and clean this up (it has to be adjusted for Healthchecker, but I can make a typed sender, and then we won't have the completely untyped thing here.

Reply:
right, maybe we have a trait like 

```rust
#[async_trait]
pub trait NetworkSender<T: Message>: Clone {
    fn send_to(
        &mut self,
        recipient: PeerId,
        protocol: ProtocolId,
        message: T,
    ) -> Result<()>;

    async fn send_rpc(&mut self, /* .. */) -> Result<T>;
}
```

Reply:
Yeah, but I don't want to do that in this PR, if I do too much, they become really long, so one piece at a time.